File: matlab/SPmakeImage.m
function [sMerge] = SPmakeImage(sMerge,indImage,indLines)

% This function generates a resampled scanning probe image with dimensions
% of imageSize, from a an array of N scan lines given in scaneLines,
% (lines specified as image rows), from an array of Nx2 origins in scanOr.
% scanDir is a 2 element vector specifying the direction of the scan.
% All arrays are stored inside struct sMerge.  ind specified update index.
% indLines is a vector of binary values specifying which lines to include.

if nargin == 2
    %     indLines = 1:size(sMerge.scanLines,2);
    indLines = true(1,size(sMerge.scanLines,2));
end

% Expand coordinates
t = repmat(1:size(sMerge.scanLines,2),[sum(indLines) 1]);
x0 = repmat(sMerge.scanOr(indLines,1,indImage),[1 size(sMerge.scanLines,2)]);
y0 = repmat(sMerge.scanOr(indLines,2,indImage),[1 size(sMerge.scanLines,2)]);
xInd = x0(:) + t(:)*sMerge.scanDir(indImage,1);
yInd = y0(:) + t(:)*sMerge.scanDir(indImage,2);

% Prevent pixels from leaving image boundaries
xInd = max(min(xInd,sMerge.imageSize(1)-1),1);
yInd = max(min(yInd,sMerge.imageSize(2)-1),1);

% Convert to bilinear interpolants and weights
xIndF = floor(xInd);
yIndF = floor(yInd);
xAll = [xIndF xIndF+1 xIndF xIndF+1];
yAll = [yIndF yIndF yIndF+1 yIndF+1];
dx = xInd-xIndF;
dy = yInd-yIndF;
w = [(1-dx).*(1-dy) dx.*(1-dy) (1-dx).*dy dx.*dy];
indAll = sub2ind(sMerge.imageSize,xAll,yAll);

% Generate image
sL = sMerge.scanLines(indLines,:,indImage);
sig = reshape(accumarray(indAll(:),[ ...
    w(:,1).*sL(:);
    w(:,2).*sL(:);
    w(:,3).*sL(:);
    w(:,4).*sL(:)],...
    [prod(sMerge.imageSize) 1]),sMerge.imageSize);
count = reshape(accumarray(indAll(:),[ ...
    w(:,1);w(:,2);w(:,3);w(:,4)],...
    [prod(sMerge.imageSize) 1]),sMerge.imageSize);

% Apply KDE
r = max(ceil(sMerge.KDEsigma*3),5);
sm = fspecial('gaussian',2*r+1,sMerge.KDEsigma);
sm = sm / sum(sm(:));
sig = conv2(sig,sm,'same');
count = conv2(count,sm,'same');
sub = count > 0;
sig(sub) = sig(sub) ./ count(sub);
sMerge.imageTransform(:,:,indImage) = sig;

% Estimate sampling density
bound = count == 0;
bound([1 end],:) = true;
bound(:,[1 end]) = true;
sMerge.imageDensity(:,:,indImage) = ...
    sin(min(bwdist(bound)/sMerge.edgeWidth,1)*pi/2).^2;


% % Plot testing
% % dens = sMerge.imageDensity(:,:,indImage);
% figure(1)
% clf
% imagesc(sig)
% % imagesc(sig.*dens + (1-dens)*mean(sig(dens>.5)))
% % imagesc(sMerge.imageTransform(:,:,indImage))
% axis equal off
% set(gca,'position',[0 0 1 1])
% % imagesc(t)
% % hold on
% % ind = 17:3213:numel(xInd);
% % scatter(yInd(ind),xInd(ind),'r.')
% % % hold off
% % % axis equal off
% % % colormap(gray(256))


end

File: scanning_drift_corr/SPmakeImage.py
"""This file contains the function SPmakeImage
"""

import numpy as np

from scanning_drift_corr.tools import distance_transform, \
    bilinear_interpolation, apply_KDE

# Developer's use, for normal usage, should be 0. A small number to
# the 'count' array to be consistent with MATLAB while checking the
# logical of implementation.
DELTA = 0

def SPmakeImage(sMerge, indImage, indLines=None):
    """
    This function generates a resampled scanning probe image with dimensions
    of imageSize, from a an array of N scan lines given in scaneLines,
    (lines specified as image rows), from an array of Nx2 origins in scanOr.
    scanDir is a 2 element vector specifying the direction of the scan.
    All arrays are stored inside struct sMerge.  ind specified update index.
    indLines is a vector of binary values specifying which lines to include.

    Parameters
    ----------
    sMerge : sMerge object
        the sMerge object.
    indImage : int
        the index of the image to be transformed.
    indLines : ndarray, optional
        an array of binary values specifying which lines to include.
        The default is None, set to use all rows.

    Returns
    -------
    sMerge : sMerge object
        the sMerge object.
    """

    # perform bilinear interpolation
    scanLines = sMerge.scanLines[indImage, ...]
    scanOr = sMerge.scanOr[indImage, ...]
    scanDir = sMerge.scanDir[indImage, :]
    imageSize = sMerge.imageSize
    sig, count = bilinear_interpolation(scanLines, scanOr, scanDir, imageSize,
                                        indLines=indLines)

    # Apply KDE
    sig = apply_KDE(sig, sMerge.KDEsigma)
    count = apply_KDE(count, sMerge.KDEsigma)

    # cheat mode!
    if DELTA:
        count += DELTA

    # the precision(?) in MATLAB sometimes results in edge value being evaluated
    # as zero while it is not (shouldn't be worried?)
    sub = count > 0
    sig[sub] /= count[sub]
    sMerge.imageTransform[indImage, ...] = sig

    # Estimate sampling density
    bound = count == 0
    bound[[0,-1], :] = True
    bound[:, [0, -1]] = True

    # MATLAB bwdist calculates 'the distance between that pixel and the
    # nearest nonzero pixel', scipy version is more conventional, which is
    # the reverse, use a wrapper to handle this
    dt = distance_transform(bound)
    dtmin = np.minimum(dt/sMerge.edgeWidth, 1)
    sMerge.imageDensity[indImage, ...] = np.sin(dtmin*np.pi/2)**2

    return sMerge


def makeImage(scanLines, scanOr, scanDir, imageSize, KDEsigma):
    """Generate the resampled image only by using data not in the sMerge
    object.

    Used in the parallel search.
    """

    # perform bilinear interpolation
    sig, count = bilinear_interpolation(scanLines, scanOr, scanDir, imageSize)

    # Apply KDE
    sig = apply_KDE(sig, KDEsigma)
    count = apply_KDE(count, KDEsigma)

    # cheat mode!
    if DELTA:
        count += DELTA

    sub = count > 0
    sig[sub] /= count[sub]

    return sig
